src/crimson/os/cyanstore/cyan_store.h

```diff
@@ -17,138 +17,277 @@
+    boost::intrusive_ptr<Collection> _get_collection(const coll_t& cid);
+  };
+
+  seastar::sharded<ShardStores> shard_stores;
 public:
+
   CyanStore(const std::string& path);
   ~CyanStore() final;
 
   seastar::future<> stop() final {
-    return seastar::now();
+    return shard_stores.stop();
   }
-  mount_ertr::future<> mount() final;
-  seastar::future<> umount() final;
+  seastar::future<> start() final;
+  mkfs_ertr::future<> mkfs(uuid_d new_osd_fsid);
 
-  mkfs_ertr::future<> mkfs(uuid_d new_osd_fsid) final;
   seastar::future<store_statfs_t> stat() const final;
+  seastar::future<std::vector<coll_t>> list_collections() final;
+
+  seastar::future<> write_meta(const std::string& key,
+		  const std::string& value) final;
+  seastar::future<std::tuple<int, std::string>>
+  read_meta(const std::string& key) final;
+  uuid_d get_fsid() const final;
+  unsigned get_max_attr_name_length() const final;
+
+
+  mount_ertr::future<> mount() final {
+    return shard_stores.invoke_on_all(
+      [](auto &local_store) {
+      return local_store.mount().handle_error(
+      crimson::stateful_ec::handle([](const auto& ec) {
+        crimson::get_logger(ceph_subsys_cyanstore).error(
+	    "error mounting cyanstore: ({}) {}",
+            ec.value(), ec.message());
+        std::exit(EXIT_FAILURE);
+      }));
+    });
+  }
+
+  seastar::future<> umount() final {
+    return shard_stores.invoke_on_all(
+      [](auto &local_store) {
+      return local_store.umount();
+    });
+  }
+
   seastar::future<struct stat> stat(
     CollectionRef c,
-    const ghobject_t& oid) final;
+    const ghobject_t& oid) final {
+    return shard_stores.local().stat(
+      c, oid);
+  }
 
   read_errorator::future<ceph::bufferlist> read(
     CollectionRef c,
     const ghobject_t& oid,
     uint64_t offset,
     size_t len,
-    uint32_t op_flags = 0) final;
+    uint32_t op_flags = 0) final {
+    return shard_stores.local().read(
+      c, oid, offset, len, op_flags);
+  }
+
   read_errorator::future<ceph::bufferlist> readv(
     CollectionRef c,
     const ghobject_t& oid,
     interval_set<uint64_t>& m,
-    uint32_t op_flags = 0) final;
+    uint32_t op_flags = 0) final {
+    return shard_stores.local().readv(
+      c, oid, m, op_flags);
+  }
 
   get_attr_errorator::future<ceph::bufferlist> get_attr(
     CollectionRef c,
     const ghobject_t& oid,
-    std::string_view name) const final;
+    std::string_view name) const final {
+    return shard_stores.local().get_attr(
+      c, oid, name);
+  }
+
   get_attrs_ertr::future<attrs_t> get_attrs(
     CollectionRef c,
-    const ghobject_t& oid);
+    const ghobject_t& oid) {
+    return shard_stores.local().get_attrs(
+      c, oid);
+  }
 
   read_errorator::future<omap_values_t> omap_get_values(
     CollectionRef c,
     const ghobject_t& oid,
-    const omap_keys_t& keys) final;
+    const omap_keys_t& keys) final {
+    return shard_stores.local().omap_get_values(
```

Comment:
The code looks good except only one question.

IIUC, for every cyanstore interface invocation, we first access the `shard_stores` field, and then call the corresponding `ShardStore`. This might mean that each `CyanStore` invocation has the possibility of causing cross-numa-node memory accesses.

If so, I wonder if this is possible: make `ShardStore` inherit from `FuturizedStore`, and feed each PG shard in the upper layer with the specific `ShardStore` as a `FuturizedStore` instance. If this is feasible, I think maybe we can avoid the above cross-numa-node mem accesses while keeping the FuturizedStore interface intact.

Am I right?

Reply:
@xxhdx1985126 osd will deliver pg to proper shard, so when each pg core call cyanstore common instance, it is already on the proper core, then in shardstore(it is sharded service), just get the current core's local instance will be OK, it will not cross-cores, if cross-cores, then it should be a bug.
For your suggestion,  I have one PR: #48540, please see the discussion, seems they don't think that is a good way. (that PR has a little bug not fix, I should let each shardstore use its own collections just like this PR I did).

Reply:
@xxhdx1985126 CyanStore is *much* simpler than what we can expect a real FuturizedStore implementation like SeaStore to need.  Accessing CyanStore::shard_stores::local() is a trivial amount of overhead compared with caching the value in ShardServices, and allows other implementations more flexibility to use something more complex than simply inheriting from seastar::sharded.

Reply:
> @xxhdx1985126 osd will deliver pg to proper shard, so when each pg core call cyanstore common instance, it is already on the proper core, then in shardstore(it is sharded service), just get the current core's local instance will be OK, it will not cross-cores, if cross-cores, then it should be a bug.
> For your suggestion, I have one PR: #48540, please see the discussion, seems they don't think that is a good way. (that PR has a little bug not fix, I should let each shardstore use its own collections just like this PR I did).

Sorry that I didn't make myself clear.

It looks to me that there's only one `FuturizedStore` instance held by the `OSD` instance across the whole osd. So, for us to call `ShardStore`'s methods, we have to first access the single `FuturizedStore` instance which is residing on one of the cores that the osd is running on(Probably core 0?). If this is correct, `FuturizedStore` instance accesses from other core are possibly `cross-core` ones. Am I missing anything?

And I was not suggesting that we put `seastar::sharded<ShardServices>` into `PGShardManager`, I was actually suggesting that we make `ShardStore` inherit from `FuturizedStore`, and make each `PerShardState` holds a dedicated `FuturizedStore` instance instead of sharing a common `FuturizedStore` instance among all the `PerShardState` instance. I think it's something like:

```
class ShardStore : public FuturizedStore {
...
};

class PerShardState {
    ....
    crimson::os::FuturizedStore store;
    ....
};
```

@liu-chunmei Does this make sense to you? Or am I missing anything?:-)

Reply:
> It looks to me that there's only one `FuturizedStore` instance held by the `OSD` instance across the whole osd. So, for us to call `ShardStore`'s methods, we have to first access the single `FuturizedStore` instance which is residing on one of the cores that the osd is running on(Probably core 0?). If this is correct, `FuturizedStore` instance accesses from other core are possibly `cross-core` ones. Am I right?

Not really.  It's a single read in an effectively global, immutable struct so the overhead should be trivial.

See my above comment.

Reply:
> Not really. It's a single read in an effectively global, immutable struct so the overhead should be trivial.

Got it.